workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/StateRegistryAggregator.kt

```diff
@@ -0,0 +1,300 @@
+      restoreFromBundle(
+        parentRegistryOwner!!.savedStateRegistry.consumeRestoredStateForKey(parentKey!!)
+      )
+    }
+  }
+
+  /**
+   * Must be called when the owning container view gets attached to the window.
+   * The owning view should find its [parentOwner] (probably via
+   * [WorkflowAndroidXSupport.stateRegistryOwnerFromViewTreeOrContext])
+   * and determine a string key unique within
+   * that parent to save and restore this class in that registry. These values will be
+   * cached in this object for [detachment][detachFromParentRegistry] later.
+   *
+   * This method will register on the [parentOwner]'s registry to save any child registries
+   * created with [installChildSavedStateRegistryControllerOn].
+   *
+   * If this object has not been restored yet, this method will start listening to
+   * the [parentOwner]'s lifecycle to know when to restore.
+   *
+   * Must be accompanied by a call to [detachFromParentRegistry] when the container
+   * view is detached.
+   *
+   * @param key an id for this [StateRegistryAggregator], uniquely identifying
+   * it in [parentOwner]. Usually generated by
+   * [createStateRegistryKeyForContainer][WorkflowAndroidXSupport.createStateRegistryKeyForContainer]
+   */
+  public fun attachToParentRegistry(
+    key: String,
+    parentOwner: SavedStateRegistryOwner
+  ) {
+    // attachToParentRegistry may be called multiple times without a matching detach in some cases,
+    // eg. when certain UI tests have failed and are being torn down. Ensure that if that happens
+    // we detach from the previous parent first.
+    detachFromParentRegistry()
+
+    this.parentRegistryOwner = parentOwner
+    this.parentKey = key
+
+    // We can only be restored once.
+    if (isRestored) return
+
+    val parentRegistry = parentOwner.savedStateRegistry
+    val parentLifecycle = parentOwner.lifecycle
+
+    // If the key is already registered, SavedStateRegistry throws an exception that doesn't provide
+    // enough information to troubleshoot, so we add some ourselves.
+    try {
+      parentRegistry.registerSavedStateProvider(key, ::saveToBundle)
+    } catch (e: IllegalArgumentException) {
+      throw IllegalArgumentException(
+        "Error registering SavedStateProvider: key \"$key\" is already in " +
+          "use on parent SavedStateRegistryOwner $parentOwner.\n" +
+          "This is most easily remedied by giving your container Screen rendering a unique " +
+          "Compatible.compatibilityKey, perhaps by wrapping it with Named.",
+        e
+      )
+    }
+
+    // Even if the parent lifecycle is in a state further than CREATED, new observers are sent all
+    // the lifecycle events required to catch the observer up to the current state, so this handles
+    // both the cases where we're ready to immediately restore, and where we have to wait.
+    parentLifecycle.addObserver(lifecycleObserver)
```

Comment:
Ahh. tada. this is the answer to my query above I guess. Good to know.

Reply:
It's SO handy. My single favorite thing about `Lifecycle`.